### PR TITLE
chore: implement BiDi sendCharacter

### PR DESCRIPTION
**Pull request by jrandolf-google**

### CODE REVIEW

File: packages/puppeteer-core/src/bidi/Input.ts [hunk 1]
```diff
@@ -415,6 +415,15 @@ export class BidiKeyboard extends Keyboard {
       ],
     });
   }
+
+  override async sendCharacter(char: string): Promise<void> {
+    if ([...char].length > 1) {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/bidi/Input.ts:422]:
```suggestion
    if (char.length > 1) {
```

File: packages/puppeteer-core/src/bidi/Input.ts [hunk 2]
```diff
@@ -415,6 +415,15 @@ export class BidiKeyboard extends Keyboard {
...
+      throw new Error('Cannot send more than 1 character.');
+    }
+    await this.#context.evaluate(async char => {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/bidi/Input.ts]: this probably needs to go via the sandbox (let's add a todo?)

File: packages/puppeteer-core/src/bidi/Input.ts (hunk 1 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/bidi/Input.ts:422]: [...char].length measures the number of UTF-32 unicode characters.
char.length measures the UTF-16 length.

**Review by Lightning00Blade — APPROVED** [review]

**Review by OrKoN — APPROVED** [review]

File: packages/puppeteer-core/src/bidi/Page.ts [hunk 3]
```diff
@@ -282,6 +290,27 @@ export class BidiPage extends Page {
...
+  /**
+   * @internal
+   */
+  async focusedFrame(): Promise<BidiFrame> {
+    using frame = await this.mainFrame()
+      .isolatedRealm()
+      .evaluateHandle(() => {
+        let frame: HTMLIFrameElement | undefined;
+        let win: Window | null = window;
+        while (win?.document.activeElement instanceof HTMLIFrameElement) {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/bidi/Page.ts:302]: this does not work for OOPIFs.

**Pull request merged**